### PR TITLE
[FEATURE] Add function remove to seqan3::configuration.

### PR DESCRIPTION
**Pull request by smehringer**
This will make it easy to remove a dynamic configuration selector (see #1580 for details on the naming) and replace it with a static configuration.

E.g.
```cpp
    seqan3::configuration user_cfg = seqan3::search_cfg::mode{seqan3::search_cfg::hit_all};

    // somewhere over the rainbow
    auto mode_selection = get<seqan3::search_cfg::mode>(user_cfg).selection;
    auto stripped_cfg = user_cfg.template remove<seqan3::search_cfg::mode>(); // remove dynamic mode
    // select static config
    if (mode_selection == seqan3::search_cfg::hit_all)
        return  | seqan3::search_cfg::hit_all;
    else if (...)
    // ...
```

### CODE REVIEW

**Comment by marehr** [comment]
We discussed in the Core-Meeting that the approach of removing seem like a good idea.

We also thought about sorting the configurations by id to have a consistent type every time.

File: include/seqan3/core/algorithm/configuration.hpp [hunk 1]
```diff
@@ -511,6 +542,27 @@ class configuration : public std::tuple<configs_t...>
...
+        {
+            return configuration<>{}; // empty configuration
+        }
+        else
+        {
+            using std::get;
+            auto [head, middle] = tuple_split<index>(static_cast<base_type>(*this));
+            auto tail = get<1>(tuple_split<1>(middle));
+
+            return make_configuration(std::tuple_cat(head, tail));
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/core/algorithm/configuration.hpp:563]: Yes all the tuple utility handles it correctly but `make_configuration` requires a non-empty tuple. I can also add the changes to `make_configuration` to handle empty tuples. Whatever is preffered.

File: include/seqan3/core/algorithm/configuration.hpp [hunk 2]
```diff
@@ -425,38 +456,40 @@ class configuration : public std::tuple<configs_t...>
     //!\}
 
     /*!\brief Creates a new configuration object by recursively adding the configs from the tuple.
-     * \tparam    tuple_t The tuple from which to create a new configuration object; must model
-                          seqan3::tuple_like and must at least have one element.
+     * \tparam    tuple_t The tuple from which to create a new configuration object; must model seqan3::tuple_like.
```
Comment (rrahn) [inline include/seqan3/core/algorithm/configuration.hpp:459]:
```suggestion
     * \tparam tuple_t The tuple from which to create a new configuration object; must model seqan3::tuple_like.
```

**Comment by smehringer** [comment]
Jenkins isn't updating itself but is also green

**Comment by smehringer** [comment]
@rrahn When you review, can you do it more thoroughly again? 
In my last commit I added more sematic stuff: 
- The functions are now `const` qualified since they return a new configuration
- The functions are now declared `[[nodiscard]]` to avoid mishandling the function because one might think `cfg.remove...` will remove something from the current `cfg` but instead, the user needs to create a new config from it!

**Comment by marehr** [comment]
>     * The functions are now `const` qualified since they return a new configuration

I also thought about this; The standard has similar concept for ranges, called `remove_copy`;

https://en.cppreference.com/w/cpp/algorithm/remove_copy

I think if you read this you will at least trip over the `_copy` part and will either think

* This removes the config from the same configuration, right? But what does the copy part mean? or
* Which copy is removed? or
* It removes something in a copy? or
* It copies and while copying it removes something?

Whichever initial interpretation you take they all seem ambiguous, so you will lookup the semantics in our documentation.

I like the description of the standard

> Copies elements from the range [first, last), to another range beginning at d_first, omitting the elements which satisfy specific criteria. Source and destination ranges cannot overlap. 

could be changed to

> Copies configs from the configuration, to a new configuration, omitting the config which shall be deleted.

(A configuration is like a range over types)

**Comment by smehringer** [comment]
So you are suggesting to rename it right? I agree that just calling it `remove` suggests a different behaviour. I'm fine with `remove_copy`, maybe also `remove_from_copy` to be more expressive but either way the brief you suggested will make it clear.

**Comment by rrahn** [comment]
I like the idea with remove_copy. Also we might need it for other functions as well. push_back for example or replace_with if needed at somepoint. So basically all interfaces need to reflect that this is not modified and a copy is returned. The thing that is a little weird is that it is a member function. Maybe we can think about making them hidden friends at some point:

```cpp
// modifiers
push_back_copy(configuration, config_element);
remove_copy<config_type>(configuration);
replace_copy<config_type>(configuration, new_config_element);

// accessors
has_element<config_type>(configuration);
get<config_type>(configuration);
get_or<config_type>(configuration, alternative);
```

This could help to avoid the relatively ugly template syntax when using the functions. This makes this class a lot more user firendly if they want to use it for their own algorithms.

**Comment by smehringer** [comment]
Does that mean this PR should already adapt the renaming or not? I'm confused because you agreed to the renaming but approved the PR.

**Comment by rrahn** [comment]
> Does that mean this PR should already adapt the renaming or not? I'm confused because you agreed to the renaming but approved the PR.

No, before we should discuss my proposal. Then we can change it alltogether. But you might want to rebase? Or can I squash merge?

**Pull request merged**